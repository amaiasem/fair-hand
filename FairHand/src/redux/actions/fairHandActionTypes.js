export default {
  LOAD_ALL_SHOPS: 'LOAD_ALL_SHOPS'
}
