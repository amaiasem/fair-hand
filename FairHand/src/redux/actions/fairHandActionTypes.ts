export default {
  LOAD_ALL_SHOPS: 'LOAD_ALL_SHOPS',
  FILTER_SHOP_BY_TYPE: 'FILTER_SHOP_BY_TYPE',
  FILTER_SHOP_BY_NAME: 'FILTER_SHOP_BY_NAME',
  GET_REVIEWS_BY_SHOP_NAME: 'GET_REVIEWS_BT_SHOP_NAME',
  GET_REVIEWS_BY_USER_NAME: 'GET_REVIEWS_BY_USER_NAME',
  USER_SIGN_IN: 'USER_SIGN_IN',
  USER_REGISTER: 'USER_REGISTER',
  USER_SIGN_OUT: 'USER_SIGN_OUT',
  ADD_REVIEW: 'ADD_REVIEW'
}
