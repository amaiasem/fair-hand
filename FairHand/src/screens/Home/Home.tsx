import React, { useEffect } from 'react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { StyleSheet, Text, View, FlatList, Image } from 'react-native'
import { TouchableOpacity } from 'react-native-gesture-handler'
import renderHeader from '../../Components/header/Header'
import { COLOR, SIZES, SHADOW2 } from '../../../constants'
import loadAllShops, { filterShopsByType, filterShopsByName } from '../../redux/actions/fairHandActionCreators'
import ShopInterface from '../../Interfaces/shopInterface'
import RenderSearchWithFilters from '../../Components/SearchWithFilters/SearchWithFilters'

const styles = StyleSheet.create({
  container: {
    marginTop: 30,
    flex: 1,
    alignItems: 'center'
  },
  containerShops: {
    paddingBottom: 200,
    marginTop: 140,
    marginBottom: 50
  },
  shopCard: {
    flex: 1,
    flexDirection: 'row',
    width: SIZES.width * 0.95,
    height: 140,
    backgroundColor: COLOR.white,
    marginBottom: 10,
    marginRight: 5,
    marginLeft: 5,
    borderRadius: 5,
    ...SHADOW2
  },
  containerImage: {
    flex: 0.5,
    justifyContent: 'center',
    alignItems: 'center',
    marginRight: 5,
    marginLeft: 10
  },
  shopImage: {
    resizeMode: 'contain',
    width: 100,
    height: 100,
    borderRadius: SIZES.cardRadius
  },
  shopInfo: {
    flex: 1,
    justifyContent: 'flex-start',
    alignItems: 'flex-start',
    padding: 5,
    marginTop: 10
  },
  shopName: {
    fontSize: SIZES.h2,
    fontWeight: '700',
    color: COLOR.orange
  },
  address: {
    fontSize: SIZES.p14
  },
  schedule: {
    fontSize: SIZES.p14,
    fontStyle: 'italic',
    color: COLOR.lightgrey
  }
})

const Home = ({ shops, filteredShops, action, navigation }:
  {shops: ShopInterface[], filteredShops: ShopInterface[], action: any, navigation: any}) => {
  useEffect(() => {
    setTimeout(() => {
      action.loadAllShops()
    }, 400)
  }, [])

  const renderItem = ({ item }: any) => (
    <TouchableOpacity
    style={styles.shopCard}
    onPress={() => navigation.navigate('Shop', { item })}
    >
      <View style= {styles.containerImage}>
          <Image
            key = {item._id}
            style= {styles.shopImage}
            source={{ uri: item.logoImage }}
          ></Image>
      </View>
      <View style={styles.shopInfo}>
        <Text style={styles.shopName}>{item.shopName}</Text>
        <Text style={styles.address}>{item.address}</Text>
        <Text style={styles.schedule}>{item.schedule}</Text>
      </View>
    </TouchableOpacity>
  )

  return (
    <View style = {styles.container}>
      {renderHeader()}
      <RenderSearchWithFilters/>
      <FlatList
        data={filteredShops}
        renderItem={renderItem}
        keyExtractor={item => item._id}
        style={styles.containerShops}
      />
    </View>
  )
}

function mapStateToProps (state: any) {
  return {
    shops: state.shopReducer.shops,
    filteredShops: state.shopReducer.filteredShops
  }
}

function mapDispatchToProps (dispatch: any) {
  return {
    action: bindActionCreators({ loadAllShops, filterShopsByType, filterShopsByName }, dispatch)
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Home)
