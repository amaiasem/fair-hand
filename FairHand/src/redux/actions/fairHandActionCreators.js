import axios from 'axios'
import fairHandActionTypes from './fairHandActionTypes'

export default function loadAllShops () {
  return async function fetchInfo (dispatch) {
    const { data } = await axios('http://192.168.0.41:5000/shops')
    dispatch({
      type: fairHandActionTypes.LOAD_ALL_SHOPS,
      data
    })
  }
}

export function filterShopsByType (shops, typeClothes) {
  const filteredShops = shops.filter(shop => shop.type.includes(typeClothes) === true)

  return {
    type: fairHandActionTypes.FILTER_SHOP_BY_TYPE,
    data: filteredShops
  }
}

export function filterShopsByName (shops, shopName) {
  const filteredShops = shops.filter(shop => shop.shopName.includes(shopName) === true)

  return {
    type: fairHandActionTypes.FILTER_SHOP_BY_NAME,
    data: filteredShops
  }
}
