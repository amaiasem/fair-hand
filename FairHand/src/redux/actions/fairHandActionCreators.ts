import axios from 'axios'
import fairHandActionTypes from './fairHandActionTypes'
import ShopInterface from '../../Interfaces/shopInterface'
import Review from '../../Interfaces/reviewInterface'

// Shop actions

export function loadAllShops () {
  return async function fetchInfo (dispatch: any) {
    const { data } = await axios.get('http://192.168.0.41:5000/shops')
    dispatch({
      type: fairHandActionTypes.LOAD_ALL_SHOPS,
      data
    })
  }
}

export function filterShopsByType (shops: ShopInterface[], typeClothes: string) {
  const filteredShops = shops.filter(shop => shop.type.includes(typeClothes) === true)

  return {
    type: fairHandActionTypes.FILTER_SHOP_BY_TYPE,
    data: filteredShops
  }
}

export function filterShopsByName (shops: ShopInterface[], shopName: string) {
  if (shopName === '') {
    return {
      type: fairHandActionTypes.FILTER_SHOP_BY_NAME,
      data: shops
    }
  } else {
    const filteredShops = shops.filter(shop => shop.shopName.includes(shopName) === true)

    return {
      type: fairHandActionTypes.FILTER_SHOP_BY_NAME,
      data: filteredShops
    }
  }
}

// Reviews actions

export function getReviewsByShopName (searchShop:string) {
  return async function fetchInfo (dispatch: any) {
    const { data } = await axios.get(`http://192.168.0.41:5000/reviews/shopName/${searchShop}`)

    dispatch({
      type: fairHandActionTypes.GET_REVIEWS_BY_SHOP_NAME,
      data
    })
  }
}

export function getReviewsByUserName (searchName:string) {
  return async function fetchInfo (dispatch: any) {
    const { data } = await axios.get(`http://192.168.0.41:5000/reviews/userName/${searchName}`)

    dispatch({
      type: fairHandActionTypes.GET_REVIEWS_BY_USER_NAME,
      data
    })
  }
}

// User actions

export function userLogin (user: object) {
  return async (dispatch: any) => {
    try {
      const { data } = await axios.post('http://192.168.0.41:5000/auth/login', user)
      dispatch({
        type: fairHandActionTypes.USER_SIGN_IN,
        data
      })
    } catch (error) {
      dispatch({
        type: fairHandActionTypes.USER_SIGN_IN,
        data: 400
      })
    }
  }
}

export function userRegister (user: object) {
  return async (dispatch: any) => {
    const { data } = await axios.post('http://192.168.0.41:5000/auth/register', user)
    if (data === 'User already exists!') {
      dispatch({
        type: fairHandActionTypes.USER_REGISTER
      })
    } else {
      dispatch({
        type: fairHandActionTypes.USER_REGISTER,
        data
      })
    }
  }
}

export function addReview (review: Review) {
  return async (dispatch: any) => {
    try {
      const { data } = await axios.post('http://192.168.0.41:5000/reviews', review)
      dispatch({
        type: fairHandActionTypes.ADD_REVIEW,
        data
      })
    } catch (error) {
      dispatch({
        type: fairHandActionTypes.ADD_REVIEW,
        data: 400
      })
    }
  }
}

export function deleteReview (allReviews: Review[], reviewID: Review) {
  return async (dispatch: any) => {
    try {
      const { data } = await axios.delete('http://192.168.0.41:5000/reviews', {
        data: {
          _id: reviewID
        }
      })
      dispatch({
        type: fairHandActionTypes.DELETE_REVIEW,
        data: { allReviews: allReviews, deleted: data }
      })
    } catch (error) {
      dispatch({
        type: fairHandActionTypes.DELETE_REVIEW,
        data: 400
      })
    }
  }
}

export function userUpdate (user: object) {
  return async (dispatch: any) => {
    const { data } = await axios.put('http://192.168.0.41:5000/user', user)
    if (data === 'Could not update the user') {
      dispatch({
        type: fairHandActionTypes.USER_UPDATE
      })
    } else {
      dispatch({
        type: fairHandActionTypes.USER_UPDATE,
        data
      })
    }
  }
}
