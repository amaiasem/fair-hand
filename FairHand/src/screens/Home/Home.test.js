import React from 'react'
import { Provider } from 'react-redux'
import renderer from 'react-test-renderer'
import { fireEvent, cleanup, act } from '@testing-library/react-native'
import Home from './Home.tsx'
// import { DATA } from '../../constants'
import configureStore from 'redux-mock-store'
import * as action from '../../redux/actions/fairHandActionCreators'
jest.mock('../../redux/actions/fairHandActionCreators')

const mockStore = configureStore([])

describe('Given a Connected React-Redux Home Component', () => {
  let store
  let component

  beforeEach(() => {
    store = mockStore({
      shopReducer: { shops: [], shop: {}, filteredShops: [] },
      reviewReducer: { reviews: [] }
    })

    afterEach(() => cleanup())

    component = renderer.create(
      <Provider store={store}>
        <Home />
      </Provider>
    )
  })

  it('It should render with given state from Redux store', () => {
    expect(component.toJSON()).toMatchSnapshot()
  })

  // describe('Given a TextInput when searching Brava', () => {
  //   it('Text input value should be Brava', async () => {
  //     const { getByTestId } = await renderer.create(
  //       <Provider store={store}>
  //         <Home />
  //       </Provider>
  //     )
  //     let newTestValue
  //     let input

  //     await act(async () => {
  //       newTestValue = 'Brava'
  //       input = getByTestId('input-shop')
  //       fireEvent.changeText(input, newTestValue)
  //     })
  //     expect(input.props.value).toBe(newTestValue)
  //   })
  // })
})
