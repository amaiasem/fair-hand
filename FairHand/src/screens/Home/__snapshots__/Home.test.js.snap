// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a Connected React-Redux Home Component It should render with given state from Redux store 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Memo(Connect(Home))
    navigation={
      Object {
        "navigate": [MockFunction],
      }
    }
  />
</Provider>
`;
